server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/NodeShutdownAllocationDecider.java

```diff
@@ -0,0 +1,121 @@
+     */
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        final SingleNodeShutdownMetadata thisNodeShutdownMetadata = getNodeShutdownMetadata(allocation.metadata(), node.nodeId());
+
+        if (thisNodeShutdownMetadata == null) {
+            // There's no shutdown metadata for this node, return yes.
+            return allocation.decision(Decision.YES, NAME, "this node is not currently shutting down");
+        }
+
+        switch (thisNodeShutdownMetadata.getType()) {
+            case REMOVE:
+                return allocation.decision(Decision.NO, NAME, "node [%s] is preparing to be removed from the cluster", node.nodeId());
+            case RESTART:
+                return allocation.decision(
+                    Decision.YES,
+                    NAME,
+                    "node [%s] is preparing to restart, but will remain in the cluster",
+                    node.nodeId()
+                );
+            default:
+                logger.error(
+                    "found unrecognized node shutdown type [{}] while deciding allocation for [{}] shard [{}][{}] on node [{}]",
+                    thisNodeShutdownMetadata.getType(),
+                    shardRouting.primary() ? "primary" : "replica",
+                    shardRouting.getIndexName(),
+                    shardRouting.getId(),
+                    node.nodeId()
+                );
+                assert false : "node shutdown type not recognized: " + thisNodeShutdownMetadata.getType();
+                return Decision.YES;
+        }
+    }
+
+    /**
+     * Applies the same rules as {@link NodeShutdownAllocationDecider#canAllocate(ShardRouting, RoutingNode, RoutingAllocation)} to
+     * determine if shards can remain on their current node.
+     */
+    @Override
+    public Decision canRemain(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        return this.canAllocate(shardRouting, node, allocation);
+    }
+
+    /**
+     * Prevents indices from being auto-expanded to nodes which are in the process of shutting down, regardless of whether they're shutting
+     * down for restart or removal.
+     */
+    @Override
+    public Decision shouldAutoExpandToNode(IndexMetadata indexMetadata, DiscoveryNode node, RoutingAllocation allocation) {
+        SingleNodeShutdownMetadata thisNodeShutdownMetadata = getNodeShutdownMetadata(allocation.metadata(), node.getId());
+
+        if (thisNodeShutdownMetadata == null) {
+            return allocation.decision(Decision.YES, NAME, "node [%s] is not preparing for removal from the cluster");
+        }
+
+        switch (thisNodeShutdownMetadata.getType()) {
+            case RESTART:
+                return allocation.decision(Decision.YES, NAME, "node [%s] is preparing to restart, but will remain in the cluster",
+                    node.getId());
```

Comment:
If we know a node is going to be restarted, should we actually expand replicas to the node? I'm trying to think of a scenario where we would *need* to expand replicas for safety if we knew the node was going to be restarted, but I haven't thought of any, right now it seems to make more sense to keep the data off if we know the node will be going away. What do you think?

Reply:
That does make sense - my thinking was that if a node is restarting, we can allow the replica expansion and it won't matter as we won't reassign the shard (once we implement that bit, anyway). I think we could go either way on this, but until/unless we find a case where it's beneficial to expand while the node is still shutting down, I'll change it to a `NO`.